core/src/cluster_info_repair_listener.rs

```diff
@@ -0,0 +1,683 @@
+use solana_metrics::datapoint;
+use solana_runtime::epoch_schedule::EpochSchedule;
+use solana_sdk::pubkey::Pubkey;
+use std::cmp::min;
+use std::collections::HashMap;
+use std::mem;
+use std::net::SocketAddr;
+use std::net::UdpSocket;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+use std::thread::{self, sleep, Builder, JoinHandle};
+use std::time::Duration;
+
+pub const REPAIRMEN_SLEEP_MILLIS: usize = 1000;
+pub const REPAIR_REDUNDANCY: usize = 3;
+pub const BLOB_SEND_SLEEP_MILLIS: usize = 2;
+pub const NUM_BUFFER_SLOTS: usize = 100;
+
+pub struct ClusterInfoRepairListener {
+    thread_hdls: Vec<JoinHandle<()>>,
+}
+
+impl ClusterInfoRepairListener {
+    pub fn new(
+        blocktree: &Arc<Blocktree>,
+        exit: &Arc<AtomicBool>,
+        cluster_info: Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: EpochSchedule,
+    ) -> Self {
+        let exit = exit.clone();
+        let blocktree = blocktree.clone();
+        let thread = Builder::new()
+            .name("solana-cluster_info_repair_listener".to_string())
+            .spawn(move || {
+                // Maps a peer to the last timestamp and root they gossiped
+                let mut peer_roots: HashMap<Pubkey, (u64, u64)> = HashMap::new();
+                let _ = Self::recv_loop(
+                    &blocktree,
+                    &mut peer_roots,
+                    exit,
+                    &cluster_info,
+                    epoch_schedule,
+                );
+            })
+            .unwrap();
+        Self {
+            thread_hdls: vec![thread],
+        }
+    }
+
+    fn recv_loop(
+        blocktree: &Blocktree,
+        peer_roots: &mut HashMap<Pubkey, (u64, u64)>,
+        exit: Arc<AtomicBool>,
+        cluster_info: &Arc<RwLock<ClusterInfo>>,
+        epoch_schedule: EpochSchedule,
+    ) -> Result<()> {
+        let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+        let my_id = cluster_info.read().unwrap().id();
+        let mut my_gossiped_root = 0;
+
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                return Ok(());
+            }
+
+            let peers = cluster_info.read().unwrap().gossip_peers();
+            let mut peers_needing_repairs: HashMap<Pubkey, EpochSlots> = HashMap::new();
+
+            // Iterate through all the known nodes in the network, looking for ones that
+            // need repairs
+            for peer in peers {
+                let last_update_ts = Self::get_last_ts(peer.id, peer_roots);
+                let my_root =
+                    Self::update_my_gossiped_root(&my_id, cluster_info, &mut my_gossiped_root);
+                {
+                    let r_cluster_info = cluster_info.read().unwrap();
+
+                    // Update our local map with the updated peers' information
+                    if let Some((peer_epoch_slots, ts)) =
+                        r_cluster_info.get_epoch_state_for_node(&peer.id, last_update_ts)
```

Comment:
you can read().unwrap() on this line and drop the lock sooner

Reply:
`get_epoch_state_for_node` returns a reference `peer_epoch_slots` to internal state, so we would have to clone it out immediately. I didn't want to copy a potentially large hashset in `peer_epoch_slots` (done a few lines later) unless it passed the check that it was needed.